GH-165: Basic Git Extension

kittaakos opened the pull request:
See #165

akosyakov commented on packages/git/src/common/model.ts:
```diff
@@ -0,0 +1,194 @@
+/**
+ * Representation of an individual file change in the working directory.
+ */
+export interface FileChange {
```
It would be nice to prefix such common names with `Git`, there is already `FileChange`

svenefftinge replied:
Or we put everything into the `Git` namespace?

akosyakov commented on packages/git/src/node/dugite-git.spec.ts:
```diff
@@ -0,0 +1,97 @@
+
+});
+
+async function createGit(fsRoot: string = ''): Promise<DugiteGit> {
```
It would be nice to avoid module functions and prefer class methods to be able to override them.

kittaakos replied:
It is a test, did you notice that? I mean I can, but does it make sense?

akosyakov replied:
I meant generally, I've seen unexported module functions in other modules.

akosyakov replied:
for tests, it is indeed not important

kittaakos replied:
Show me, and we replace it. I personally hate writing `this.` when it does nothing with the class' state at all.

akosyakov replied:
it should be another class then

akosyakov replied:
would be even longer though :) I understand your frustration, but we don't have dependency injection for modules, but for classes

kittaakos replied:
Make sense.

akosyakov commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -0,0 +1,210 @@
+export class DugiteGit implements Git {
+
+    constructor(
+        @inject(WorkspaceServer) private readonly workspace: WorkspaceServer) {
```
it would be nice to prefer `protected` over `private` to allow subclasses to override and access the state.

kittaakos replied:
If you prefer that, we use protected then.

akosyakov replied:
it is just a pain, for example, to work around some bugs in JDT because everything is private.

svenefftinge commented on packages/git/src/node/git-repository-locator.ts:
```diff
@@ -0,0 +1,73 @@
+ *
+ * @param path the FS path of the root to start the discovery.
+ */
+export async function locateRepositories(path: string): Promise<Repository[]> {
```
`async` is not used in the code

svenefftinge commented on packages/git/src/node/git-repository-locator.ts:
```diff
@@ -0,0 +1,73 @@
+    return !parts[0].length ? parts.slice(1) : parts;
+}
+
+// function getOrigin(repository, cb) {
```
remove dead code

svenefftinge commented on packages/git/src/node/git-repository-locator.ts:
```diff
@@ -0,0 +1,73 @@
+
+async function splitPath(path: string): Promise<string[]> {
+    const parts = path.split(/(\/|\\)/);
+    if (!parts.length) {
```
I'd prefer being more explicit about length === 0.

kittaakos replied:
That is bad anyway. We need to call `path.dirname()` twice.

svenefftinge commented on packages/git/src/node/git-repository-locator.ts:
```diff
@@ -0,0 +1,73 @@
+    if (!parts.length) {
+        return parts;
+    }
+    // When the `path` starts with a slash, the the first part is empty string.
```
the the

kittaakos commented on packages/git/src/browser/git-command.ts:
```diff
@@ -0,0 +1,114 @@
+        id: 'git.status',
+        label: 'Print Git Status'
+    };
+    export const REPOSITORIES = {
```
It was just a placeholder command to see if the interface can be called from the browser. We will get rid of this for sure.

kittaakos commented on packages/git/src/browser/git-keybinding.ts:
```diff
@@ -0,0 +1,46 @@
+import { KeybindingContext, Keybinding, KeybindingContribution, KeybindingRegistry, KeyCode, Key, Modifier } from "@theia/core/lib/common";
+
+@injectable()
+export class GitKeybindingContext implements KeybindingContext {
```
Will be removed.

svenefftinge commented on packages/git/src/browser/git-resource.ts:
```diff
@@ -7,19 +7,26 @@
 
 export class GitResource implements Resource {
+
+    protected readonly toDispose = new DisposableCollection();
```
can we remove it? doesn't seem to be used

svenefftinge commented on packages/git/src/browser/git-widget.ts:
```diff
@@ -207,11 +209,30 @@ export class GitWidget extends VirtualWidget {
-                ressource
-                    .then(r => r.readContents())
-                    .then(content => console.log(content));
+                // try {
```
code should be deleted

svenefftinge commented on packages/monaco/src/browser/monaco-common-editor.ts:
```diff
@@ -0,0 +1,235 @@
+/*
```
module can be removed, since it is not used.

jbicker commented on packages/git/src/browser/git-resource.ts:
```diff
@@ -0,0 +1,69 @@
+            localUri = new URI(this.repos[idx].localUri);
+            // make sure that localUri of repo has no scheme.
+            localUriStr = localUri.withoutScheme().toString();
+            if (dirStr.toString().endsWith(localUriStr)) {
```
No. If we have nested repositories I want to get the repository which is next to the given file uri. 
If I would use 'startsWith' this function would probably return the "parent"-repository.

svenefftinge commented:
<img width="767" alt="screen shot 2017-09-26 at 15 20 46" src="https://user-images.githubusercontent.com/372735/30862428-4dccd10a-a2ce-11e7-870f-eb76ebdce1dd.png">

kittaakos commented on packages/git/src/browser/style/index.css:
```diff
@@ -0,0 +1,153 @@
+/*
```
@jbicker, I was unable to figure out where do we set the `id` of the left panel, so I just add this as a comment. The problem is the following: both the `Files` and the `Git` has the same `files` ID. So the browser tests are broken.

kittaakos replied:
Of course, I was wrong. I have figured out what was the problem and it is unrelated to the `id`. The ID is just fine.

svenefftinge commented:
Merge conflicts are shown twice:
<img width="381" alt="screen shot 2017-09-27 at 10 33 15" src="https://user-images.githubusercontent.com/372735/30903703-ecd92db0-a36f-11e7-86d5-11fbcf12d2b7.png">

svenefftinge commented:
Sometimes when double-clicking on a change a diff editor gets opened twice (I will take care of that).

svenefftinge commented:
On huge repositories, the initial load sometimes has all changes twice:

<img width="268" alt="screen shot 2017-09-27 at 13 10 29" src="https://user-images.githubusercontent.com/372735/30910515-5d53ccfc-a385-11e7-8371-e1fdf3f74d8f.png">

svenefftinge commented:
Also on huge repos, the initial status request takes multiple seconds.

kittaakos commented:
> On huge repositories, the initial load sometimes has all changes twice:

I will take care of this.

kittaakos commented:
Open diff editor, close, open again. A broken editor is shown with this in the browser:
```
6:37:06.690 console-window.ts:40 TypeScript: [Error - 4:37:06 PM] Server initialization failed.
16:37:06.690 console-window.ts:40 TypeScript: Error: Connection got disposed.
    at Object.dispose (http://localhost:3000/bundle.js:6791:25)
    at http://localhost:3000/bundle.js:64956:56
    at CallbackList.invoke (http://localhost:3000/bundle.js:12821:39)
    at Emitter.fire (http://localhost:3000/bundle.js:12885:36)
    at closeHandler (http://localhost:3000/bundle.js:6187:26)
    at CallbackList.invoke (http://localhost:3000/bundle.js:12821:39)
    at Emitter.fire (http://localhost:3000/bundle.js:12885:36)
    at WebSocketMessageReader.AbstractMessageReader.fireClose (http://localhost:3000/bundle.js:22284:27)
    at WebSocketMessageReader.fireClose (http://localhost:3000/bundle.js:33595:40)
    at http://localhost:3000/bundle.js:33552:19
16:37:06.691 main.js:129 Uncaught (in promise) Error: Connection is disposed.
    at new ConnectionError (main.js:129)
    at throwIfClosedOrDisposed (main.js:607)
    at Object.sendRequest (main.js:706)
    at Object.shutdown (connection.js:20)
    at resolveConnection.then.connection (base.js:845)
    at <anonymous>
ConnectionError @ main.js:129
throwIfClosedOrDisposed @ main.js:607
sendRequest @ main.js:706
shutdown @ connection.js:20
resolveConnection.then.connection @ base.js:845
Promise resolved (async)
stop @ base.js:844
connection.initialize.then @ base.js:831
Promise rejected (async)
initialize @ base.js:747
resolveConnection.then @ base.js:657
Promise resolved (async)
start @ base.js:624
handleConnectionClosed @ base.js:919
CallbackList.invoke @ events.js:71
Emitter.fire @ events.js:135
closeHandler @ main.js:212
CallbackList.invoke @ events.js:71
Emitter.fire @ events.js:135
AbstractMessageReader.fireClose @ messageReader.js:126
WebSocketMessageReader.fireClose @ reader.ts:72
(anonymous) @ reader.ts:31
webSocket.onclose @ connection.ts:32
16:37:06.691 main.js:816 Uncaught (in promise) Error: Connection got disposed.
    at Object.dispose (main.js:816)
    at connection.ts:14
    at CallbackList.invoke (events.js:71)
    at Emitter.fire (events.js:135)
    at closeHandler (main.js:212)
    at CallbackList.invoke (events.js:71)
    at Emitter.fire (events.js:135)
    at WebSocketMessageReader.AbstractMessageReader.fireClose (messageReader.js:126)
    at WebSocketMessageReader.fireClose (reader.ts:72)
    at reader.ts:31
```

svenefftinge commented:
Open editor for added file fails:
2bundle.js:32746 Uncaught (in promise) Error: Request show failed with message: Unable to find path to repository on disk.
    at new ResponseError (bundle.js:32746)
    at handleResponse (bundle.js:6396)
    at processMessageQueue (bundle.js:6224)
    at bundle.js:6208
    at run (bundle.js:63088)
    at runIfPresent (bundle.js:63117)
    at onGlobalMessage (bundle.js:63157)

kittaakos commented:
I do not know if it is a side-effect of something else but I had this in my console, when I have tried to open an untracked file in a "diff" editor:
```
index.js:380 TypeError: Cannot read property 'style' of null
    at t.layout (editor.main.js:31)
    at t.layout (editor.main.js:31)
    at t._doLayout (editor.main.js:31)
    at t._measureDomElement (editor.main.js:31)
    at t.layout (editor.main.js:31)
    at MonacoDiffEditor.webpackJsonp.524.MonacoDiffEditor.resize (monaco-diff-editor.ts:78)
    at MonacoDiffEditor.webpackJsonp.505.MonacoEditor.setSize (monaco-editor.ts:232)
    at EditorWidget.onResize (editor-widget.ts:44)
    at EditorWidget.Widget.processMessage (widget.js:485)
    at invokeHandler (index.js:433)
```

svenefftinge commented:
> Open diff editor, close, open again. A broken editor is shown with this in the browser:

found the issue, fix is on its way

kittaakos commented:
Editor (not the diff one) content does not get updated after discarding all the changes:
 - Open a file in the editor.
 - Change the file content. (You do not have to save, you know...)
 - Leave the file opened in the editor.
 - Go to Git tab, discard changes.
 - File restored to its HEAD state.
 - The opened editor still shows the modified content.

kittaakos commented:
I will look into the windows build issue and I try to fix it.

kittaakos commented on packages/git/src/common/git-watcher.ts:
```diff
@@ -0,0 +1,188 @@
+    /**
+     * The old `Git` source. Might be missing.
+     */
+    readonly oldSource?: Repository;
```
Sure, we have to listen to two kinds of events: 
 - A new repository appears or gets removed from the workspace.
 - The other type of the event is when the status has changed in a repository.

Both of the event types is a subtype of the `GitEvent` type. Since a repository "deletion" has to be distinguished from a repository "creation" event, we have the source, which stands for the creation event, and we have the "oldSource" which is for the deletion.

Of course, we can have two different types and have the union type of them, I did not want to do that on purpose but we can change.

svenefftinge commented:
With the current state I get, the following on commit:
```
messages.js:46 Uncaught (in promise) Error: Request commit failed with message: Unable to find path to repository on disk.
    at new ResponseError (messages.js:46)
    at handleResponse (main.js:421)
    at processMessageQueue (main.js:249)
    at main.js:233
    at run (setImmediate.js:40)
    at runIfPresent (setImmediate.js:69)
    at onGlobalMessage (setImmediate.js:109)
```

svenefftinge commented on packages/git/src/browser/git-widget.ts:
```diff
@@ -193,10 +193,8 @@ export class GitWidget extends VirtualWidget {
-                        branch: this.status.branch,
-                        paths: change.uri
-                    });
+                    const options: Git.Options.Checkout.WorkingTreeFile = { paths: change.uri };
```
Why do I not need to specify the branch from which I want to checkout? AFAIU the HEAD will be detached when I am in a conflicting rebase situation.

kittaakos replied:
Then please use `treeish`. AFAIK, `branch` is a `treeish`.

```
git checkout [-f|--ours|--theirs|-m|--conflict=<style>] [<tree-ish>] [--] <paths>…​
git checkout [-p|--patch] [<tree-ish>] [--] [<paths>…​]
```

kittaakos commented on packages/monaco/src/browser/monaco-editor-provider.ts:
```diff
@@ -123,7 +164,7 @@ export class MonacoEditorProvider {
     }
 
-    protected installQuickOpenService(editor: MonacoEditor): void {
+    protected installQuickOpenService(editor: MonacoEditor | MonacoDiffEditor): void {
```
Why is that? As I see it is a subclass of the `MonacoEditor`. Is it just some leftover code?

kittaakos commented on packages/git/src/node/dugite-git-watcher.ts:
```diff
@@ -0,0 +1,66 @@
+        const timer = setInterval(async () => {
+            const status = await this.git.status(repository);
+            const oldStatus = this.status.get(repository);
+            if (this.client && !WorkingDirectoryStatus.equals(status, oldStatus)) {
```
Here.

kittaakos commented:
I got this when cloned a repository inside my workspace and double clicked on the unstaged resource in the Git tab:
```
opener-service.ts:6 Uncaught (in promise) A resource provider for 'file:///Users/akos.kitta/git/theia/examples/browser/xtext-core/' is not registered.
```

kittaakos commented:
I would like to see tooltips for the Git integration toolbar items:
<img width="96" alt="screen shot 2017-09-28 at 18 07 14" src="https://user-images.githubusercontent.com/1405703/30977489-f16acf7e-a477-11e7-9aae-76f2f19952f3.png">

Could be a Chrome issue, but I cannot see anything at all.

kittaakos commented:
I was just opening files to have the diff editor when I had this in my console:
```
JavaScript: Error: Connection got disposed.
    at Object.dispose (http://localhost:3000/bundle.js:6792:25)
    at http://localhost:3000/bundle.js:64879:56
    at CallbackList.invoke (http://localhost:3000/bundle.js:12822:39)
    at Emitter.fire (http://localhost:3000/bundle.js:12886:36)
    at closeHandler (http://localhost:3000/bundle.js:6188:26)
    at CallbackList.invoke (http://localhost:3000/bundle.js:12822:39)
    at Emitter.fire (http://localhost:3000/bundle.js:12886:36)
    at WebSocketMessageReader.AbstractMessageReader.fireClose (http://localhost:3000/bundle.js:22292:27)
    at WebSocketMessageReader.fireClose (http://localhost:3000/bundle.js:33609:40)
    at http://localhost:3000/bundle.js:33566:19
18:12:48.235 main.js:129 Uncaught (in promise) Error: Connection is disposed.
    at new ConnectionError (main.js:129)
    at throwIfClosedOrDisposed (main.js:607)
    at Object.sendRequest (main.js:706)
    at Object.shutdown (connection.js:20)
    at resolveConnection.then.connection (base.js:845)
    at <anonymous>
ConnectionError @ main.js:129
throwIfClosedOrDisposed @ main.js:607
sendRequest @ main.js:706
shutdown @ connection.js:20
resolveConnection.then.connection @ base.js:845
Promise resolved (async)
stop @ base.js:844
connection.initialize.then @ base.js:831
Promise rejected (async)
initialize @ base.js:747
resolveConnection.then @ base.js:657
Promise resolved (async)
start @ base.js:624
handleConnectionClosed @ base.js:919
CallbackList.invoke @ events.js:71
Emitter.fire @ events.js:135
closeHandler @ main.js:212
CallbackList.invoke @ events.js:71
Emitter.fire @ events.js:135
AbstractMessageReader.fireClose @ messageReader.js:126
WebSocketMessageReader.fireClose @ reader.ts:72
(anonymous) @ reader.ts:31
webSocket.onclose @ connection.ts:32
18:12:48.235 main.js:816 Uncaught (in promise) Error: Connection got disposed.
    at Object.dispose (main.js:816)
    at connection.ts:14
    at CallbackList.invoke (events.js:71)
    at Emitter.fire (events.js:135)
    at closeHandler (main.js:212)
    at CallbackList.invoke (events.js:71)
    at Emitter.fire (events.js:135)
    at WebSocketMessageReader.AbstractMessageReader.fireClose (messageReader.js:126)
    at WebSocketMessageReader.fireClose (reader.ts:72)
    at reader.ts:31
dispose @ main.js:816
(anonymous) @ connection.ts:14
CallbackList.invoke @ events.js:71
Emitter.fire @ events.js:135
closeHandler @ main.js:212
CallbackList.invoke @ events.js:71
Emitter.fire @ events.js:135
AbstractMessageReader.fireClose @ messageReader.js:126
WebSocketMessageReader.fireClose @ reader.ts:72
(anonymous) @ reader.ts:31
webSocket.onclose @ connection.ts:32
Promise rejected (async)
resolveConnection.then.connection @ base.js:845
Promise resolved (async)
stop @ base.js:844
connection.initialize.then @ base.js:831
Promise rejected (async)
initialize @ base.js:747
resolveConnection.then @ base.js:657
Promise resolved (async)
start @ base.js:624
BaseLanguageClientContribution.activate @ language-client-contribution.ts:62
(anonymous) @ languages-frontend-contribution.ts:24
Promise resolved (async)
_loop_1 @ languages-frontend-contribution.ts:23
LanguagesFrontendContribution.onStart @ languages-frontend-contribution.ts:22
FrontendApplication.startContributions @ frontend-application.ts:110
FrontendApplication.start @ frontend-application.ts:67
start @ index.js:22
Promise resolved (async)
Object.defineProperty.value @ index.js:42
__webpack_require__ @ bootstrap 1f7454f1429b04c95a79:688
Object.defineProperty.value @ bootstrap 1f7454f1429b04c95a79:789
(anonymous) @ bootstrap 1f7454f1429b04c95a79:789
```

kittaakos commented:
@hexa00, we consider this PR ready for review. Thank you for looking into the changes!

akosyakov commented on packages/editor/src/browser/editor-utility.ts:
```diff
@@ -0,0 +1,45 @@
+
+import URI from "@theia/core/lib/common/uri";
+
+export interface DiffUris {
```
the file name should be `diff-uris.ts` as `DiffUris` is the main concept

akosyakov commented on packages/editor/src/browser/editor-utility.ts:
```diff
@@ -0,0 +1,45 @@
+    }
+}
+
+export namespace DiffUriHelper {
```
I think it should be merged into `DiffUris` namespace.

akosyakov commented on packages/editor/src/browser/editor-utility.ts:
```diff
@@ -0,0 +1,45 @@
+            modified: modified.toString()
+        };
+
+        const diffUriStr = encodeURI(JSON.stringify(diffUris));
```
`encodeURI` is not necessary `Uri` is responsible for encoding:
https://github.com/Microsoft/vscode-uri/blob/2255aa25f205d1364666a6b6f9dcc51e46d403bf/lib/index.ts#L374-L376

akosyakov commented on packages/editor/src/browser/editor-utility.ts:
```diff
@@ -0,0 +1,45 @@
+        if (uri.scheme !== 'diff') {
+            throw ('URI must have scheme "diff".');
+        }
+        const diffUris = JSON.parse(decodeURI(uri.fragment));
```
`decodeURI` is not necessary `Uri` is responsible for decoding:
https://github.com/Microsoft/vscode-uri/blob/2255aa25f205d1364666a6b6f9dcc51e46d403bf/lib/index.ts#L207

hexa00 commented:
@kittaakos I'm starting to review this now...

hexa00 commented:
I have not looked at the code yet but here are my notes just playing around with it a few minutes.

Usability notes:

 - What is refresh ?
 - Clicking on a file in git doesn't open the editor
 - Git commands are not selectable from the command palette
 - Is there any keybindings for git ?
 - The current branch should be shown somewhere especially since it
   affects the pull/push operations.
 - It should be more clear that the drop down selection is selecting a
   repo or it should just be assumed from the navigator ?
 - You should be able to checkout another branch from that too like
   you can switch repo.
 - The icon for commit is way too small.. consider something like:
http://blog.atom.io/img/posts/github-package-git.png ?
 - Should git status of the file be reflected in the navigator ?
 - It would greatly help to have git log shown below the commit dialog
to know where you are too.

Operations:

 - Operations should be logged somewhere.
 - Using push/pull without any origin defined just does noting, no errors
   or logs.
 - Push/Pull/Merge should be interactive... ask the user what needs to be
   done and present defaults to be confirmed.
 - Using `--set-upstream` in git push changes the config as a side
   effect, the user git config should not be changed.
 - Fetching should not use `--prune`
 - Pull should not use `--no-rebase` why is it?
 - What is git merge doing ? clicking on it I can't see any logs and
   usually you want to select a branch to merge.. how is that done?
 - No commit with amend?

svenefftinge commented:
@hexa00 thanks for the feedback, I agree with most of your comments. But please keep in mind that this PR is about 'BASIC' git support and focuses on delivering an implementation of the description in #165. We should create follow-up tickets and smaller! PRs for additional features like amend and so on. Could you update your list and create tickets for those that would be out-of-scope for this first PR? The rest of your items can than handled in the PR.

hexa00 commented:
@svenefftinge I agree, I think we should update that list with @marcdumais-work and possibly others as I'm not 100% sure what should be part of it.

From my view:

For sure from #165 missing is:
 - Double-clicking a file entry will open a diff editor showing the changes
 - The diff editor allows to further modify the local file copy.

I also think my comments about the interactivity of the commands need to be addressed.  Maybe I missed something there... e.g: I'm puzzled by merge without any input by the user? 

Also I think we can't do without seeing the current branch written somewhere.

I'm open to discuss the rest.

hexa00 commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -0,0 +1,289 @@
+        return FileUri.fsPath(uri);
+    }
+
+    private getUri(path: string): string {
```
this doesn't return a URI.. but is getUri... confusing a bit?

svenefftinge replied:
It does return a uri as a string

hexa00 commented on packages/workspace/src/browser/workspace-commands.ts:
```diff
@@ -221,17 +221,20 @@ export class FileSystemCommandHandler implements CommandHandler {
 
     execute(...args: any[]): object | undefined {
-        const uri = this.getUri();
+        const uri = this.getUri(...args);
```
Is this used ? same for the other commands.. I don't see a uri in arg 0 anywhere?

svenefftinge replied:
It is in git-widget.ts:222.

hexa00 replied:
Ha OK thx

hexa00 commented:
BTW after all I was able to double click on the file and open the editor, seems like you really have to click on the filename, clicking left of the filename but still on the same line didn't work.

Also I saw a diff editor now in the code but I'm not sure how to trigger it ?

epatpol commented on doc/Internals.md:
```diff
@@ -347,7 +347,7 @@ container.load(frontendLanguagesModule);
 ## Complete example
 
-If you wish to see the complete implementation of what I refered too in
+If you wish to see the complete implementation of what I referred too in
```
I think for small typos like this we could push directly to master from the GH markdown editor (for those that can) so as to not clutter big PRs even more.

hexa00 replied:
Fixes that should be fixed on master in general  should be there asap to avoid duplicating work especially on long running PRs like this one

svenefftinge commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -0,0 +1,289 @@
+        }
+    }
+
+    private getFsPath(repository: Repository | string): string {
```
`Repository` is intentionally not an OO kind of thing, because it is part of the protocol.

hexa00 replied:
Ha OK makes more sense

svenefftinge commented on packages/git/src/common/git.ts:
```diff
@@ -0,0 +1,507 @@
+        Git.Options.Branch.Create |
+        Git.Options.Branch.Rename |
+        Git.Options.Branch.Delete): Promise<void | undefined | string | string[]>;
+
```
We designed it to be close to the git cli, as we can expect developers to be familiar with that.

hexa00 replied:
Yes I understand but for the API having signatures like:

```typescript
async branch(repository: Repository,
        options: Git.Options.Branch.List |
            Git.Options.Branch.Create |
            Git.Options.Branch.Rename |
            Git.Options.Branch.Delete): Promise<void | undefined | string | string[]> {
```

Really make it hard to undestand to a developer what is expected from the function.  While that's fine for command line I think for an API it's asking for trouble.

hexa00 replied:
Think about error handling too if you do:
```
try {
  branch (...
}
catch (error) {

}
```

You will have to document every variation of the errors for each command in the same function.

kittaakos replied:
We are planning to align the API with the corresponding git command as stated before. See: [`git branch`](https://git-scm.com/docs/git-branch). If it turns out that it is super inconvenient from the client's perspective, we can change the API. One can easily create a ticket with problematic real use-cases and examples. Within this ticket, one can even propose a better alternative API. Thanks for your understanding!

svenefftinge commented on packages/core/src/browser/widget-manager.ts:
```diff
@@ -52,6 +52,7 @@ export class WidgetManager {
 
     private _cachedfactories: Map<string, WidgetFactory>;
     private widgets = new Map<string, Widget>();
+    private widgetPromises = new Map<string, Promise<Widget>>();
```
It waits for the resolution, so subsequent calls with the same arguments won't trigger another factory call.

hexa00 replied:
Isn't this true also for:
```
widget = await factory.createWidget(options);
``` ?
I don't get why creating an array ?

hexa00 replied:
Also seems to me that both versions of the code will trigger multiple factory calls... 
Seems like something would need to check if the promise already exists in the array before the factory call... ?

hexa00 replied:
@svenefftinge ?

svenefftinge replied:
Not sure, I think it's all in the code. Di you forget that js engines run single threaded?

svenefftinge commented on packages/git/package.json:
```diff
@@ -0,0 +1,55 @@
+    "abs": "^1.3.8",
+    "dugite-extra": "0.0.1-alpha.10",
+    "findit2": "^2.2.3",
+    "inversify": "^4.2.0"
```
That's still an open discussion, but I will remove it for now.

svenefftinge commented on packages/git/src/browser/git-command.ts:
```diff
@@ -0,0 +1,91 @@
+export namespace GIT_COMMANDS {
+    export const FETCH = {
+        id: 'git.fetch',
+        label: 'Fetch'
```
re: keybinding: No, I don't think so

svenefftinge commented on packages/git/src/browser/git-repository-provider.ts:
```diff
@@ -0,0 +1,32 @@
+        }
+    }
+
+    select(localUri: string) {
```
a uri as a  astring

svenefftinge commented on packages/git/src/browser/git-resource.ts:
```diff
@@ -0,0 +1,76 @@
+            this.repos = await this.git.repositories();
+        }
+        if (this.repos.length > 1) {
+            sortedRepos = this.repos.sort((a, b) => b.localUri.length - a.localUri.length);
```
no it's sported by length to find the closest repos for a given uri (i.e. longest prefix)

svenefftinge commented on packages/git/src/browser/git-widget.ts:
```diff
@@ -0,0 +1,353 @@
+
+    protected getRepositoryRelativePath(absPath: string) {
+        const pathToReplace = new URI(this.repository.localUri).withoutScheme().toString();
+        return absPath.replace(pathToReplace, '').replace(/^\//, '');
```
It's really just super simply trimming logic for UI. I don't want to expose it to one of our main concepts as API.

svenefftinge commented on packages/core/src/node/file-uri.ts:
```diff
@@ -23,8 +23,12 @@ export namespace FileUri {
-        return (uri as any).codeUri.fsPath;
+    export function fsPath(uri: URI | string): string {
+        if (typeof uri === 'string') {
+            return fsPath(new URI(uri));
```
no, because it is a uri not a path.

hexa00 replied:
Yeach I guess I find uri:string, path:string -> URI quite confusing... but maybe it's just me

svenefftinge commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -0,0 +1,289 @@
+import { FileUri } from '@theia/core/lib/node/file-uri';
+import { clone } from 'dugite-extra/lib/command/clone';
+import { fetch } from 'dugite-extra/lib/command/fetch';
+import { merge } from 'dugite-extra/lib/command/merge';
```
Maybe, I think that's a matter of taste.

svenefftinge commented:
> BTW after all I was able to click on the file and open the editor, seems like you really have to click on the filename, clicking right of the filename but still on the same line didn't work.

I fixed that.

> Also I saw a diff editor now in the code but I'm not sure how to trigger it ?

You can open a diff editor on a change. Of course, it doesn't open a diff editor if it is a new file, since comparing with an empty buffer doesn't make sense.

hexa00 commented:
@svenefftinge Should the diff work on a conflict file too ? I had tried that but got pure text...

But comparing with an empty buffesr does make sense, usually I will see like
a/newfile
/b/empty
+ 
+ newline 

etc...

hexa00 commented:
I retested and I was albe to trigger the diff editor by doing a change and clicking on the modified file in the git extension tab.

I was also able to modify the file from the diff editor.

It looks good :) Only thing I would change maybe is to show trailing whitespaces in a different color or enable whitespaces viewing... but that can wait.

svenefftinge commented:
> What is refresh ?

It refreshes the view, pulling the latest changes from git status.

> Clicking on a file in git doesn't open the editor

Fixed it in latest commit.

> Some commands are not selectable from the command palette like commit ?

Yes, because they need context. It doesn't make sense to expose them through the command palette.

> Command palette commands should be prefixed like: Git: Merge

done

> Is there any keybindings for git ?

no

> The current branch should be shown somewhere especially since it affects the pull/push operations.

Depends on status bar (#600).

> It should be more clear that the drop down selection is selecting a
repo or it should just be assumed from the navigator ?

Not sure, I would prefer to leave out any unnecessary noise in the UI.

> You should be able to checkout another branch from that too like you can switch repo.

I wouldn't overload that drop down, but rather use the status bar for that. Should be discussed in a separate ticket or maybe #600.

> The icon for commit is way too small.. consider something like: http://blog.atom.io/img/posts/github-package-git.png ?

I don't agree.

> Should git status of the file be reflected in the navigator ?

Would be nice. Create a ticket?

> It would greatly help to have git log shown below the commit dialog
to know where you are too.

Could you elaborate on that?

> Operations should be logged somewhere.

Agreed : #601

> Using push/pull without any origin defined just does noting, no errors
or logs.

Should for now show a message to the user.

> Push/Pull/Merge should be interactive... ask the user what needs to be
done and present defaults to be confirmed.

We can have additional `Push...` resp. `Pull...` operations (like in vscode).

> Using --set-upstream in git push changes the config as a side effect, the user git config should not be changed. 
> Fetching should not use --prune

Changes for dugite-extra. @kittaakos could you take care?

> Pull should not use --no-rebase why is it?

That's the default behavior of `git pull`.

> What is git merge doing ? clicking on it I can't see any logs and usually you want to select a branch to merge.. how is that done?

Right, merge doesn't make sense like this at all. 
I will remove it and we will take care of that in #583 in an additional PR.

> No commit with amend?

See #602

kittaakos commented:
> Changes for dugite-extra. @kittaakos could you take care?

See:
https://github.com/TypeFox/dugite-extra/issues/2
https://github.com/TypeFox/dugite-extra/issues/3

hexa00 commented:
>It refreshes the view, pulling the latest changes from git status

Isn't that done by polling each second ?

>>  It should be more clear that the drop down selection is selecting a
>> repo or it should just be assumed from the navigator ?

>Not sure, I would prefer to leave out any unnecessary noise in the UI.

How about using the navigator ? This would remove UI.

>The icon for commit is way too small.. consider something like: http://blog.atom.io/img/posts/github-package-git.png ?

Even vscode is 50% larger.


>> It would greatly help to have git log shown below the commit dialog
> >  to know where you are too.

>Could you elaborate on that?

I would have like:

[ commit message form]

(master)(983aad) Added feature x
(master)(983afa) Added feature y
(master)(ads832) inital commit

So you know where you are, this is especially useful while doing an interactive rebase.
But we can leave that for later.

Thanks for creating tickets on the other issues.

hexa00 commented:
> >   Pull should not use --no-rebase why is it?

> That's the default behavior of git pull.

It is not --no-rebase overides a previous --rebase 

Also usually you don't have to specify an option to use the default...

kittaakos commented:
> It is not --no-rebase overides a previous --rebase
> 
> Also usually you don't have to specify an option to use the default...

https://github.com/TypeFox/dugite-extra/issues/4

kittaakos commented:
I have updated the `dugite-extra` library based on your feedback, @hexa00. Thanks for taking the time looking into that as well! A new version has been published to NPM. I have also rebased the branch from the current state of the master.

hexa00 commented:
@kittaakos Great thanks!

svenefftinge commented:
Thanks @kittaakos! 
@hexa00 is it good to merge, then?

hexa00 commented:
@svenefftinge there is still comments in the code review left unawnsered

I was also planning to test it on windows today, unless @kittaakos where you albe to do that after all ? 
Last comments I got was that this was blocking the PR

kittaakos commented:
> I was also planning to test it on windows today, unless @kittaakos where you albe to do that after all ?

Go ahead with the testing. I still have the problem with the terminal.

kittaakos commented:
> code review left unawnsered

Could you please collect those here with the links? Thanks!

hexa00 commented:
https://github.com/theia-ide/theia/pull/392#discussion_r142654060
https://github.com/theia-ide/theia/pull/392#discussion_r142655010
https://github.com/theia-ide/theia/pull/392#discussion_r142129593
https://github.com/theia-ide/theia/pull/392#discussion_r142137185

kittaakos commented on packages/git/src/common/model.ts:
```diff
@@ -0,0 +1,111 @@
+    'Deleted',
+    'Renamed',
+    'Conflicted',
+    'Copied'
```
> I wonder that this is ? The same file in two different places ?
How is git reporting this?

Sorry, I do not understand this question. Are you asking why do we have the `Copied` status?

hexa00 replied:
Yes I wonder what that is ?

kittaakos replied:
Please reference the [documentation](https://git-scm.com/docs/git-status). `C` indicates the `Copied` status with the `--porcelain=2` option. We wanted to handle all possible cases.

hexa00 replied:
OK

svenefftinge commented on packages/core/src/browser/widget-manager.ts:
```diff
@@ -73,18 +74,21 @@ export class WidgetManager {
     async getOrCreateWidget<T extends Widget>(factoryId: string, options?: any): Promise<T> {
         const key = this.toKey({ factoryId, options });
-        const existingWidget = this.widgets.get(key);
+        const existingWidget = this.widgetPromises.get(key);
```
@hexa00 here we check whether a request exists

hexa00 replied:
Haaa someone I missed that piece of code , thx! :)

svenefftinge commented on packages/core/src/browser/widget-manager.ts:
```diff
@@ -73,18 +74,21 @@ export class WidgetManager {
         }
-        const widget = await factory.createWidget(options);
+        const widgetPromise = factory.createWidget(options);
+        this.widgetPromises.set(key, widgetPromise);
```
here we update the map, to ensure it is not called twice

svenefftinge commented on packages/core/src/browser/widget-manager.ts:
```diff
@@ -73,18 +74,21 @@ export class WidgetManager {
-        const widget = await factory.createWidget(options);
+        const widgetPromise = factory.createWidget(options);
+        this.widgetPromises.set(key, widgetPromise);
+        const widget = await widgetPromise;
```
here async part starts, but widgetPromises are updated.

hexa00 replied:
K makes sense now thx

kittaakos commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -0,0 +1,289 @@
+        return stage(this.getFsPath(repository), paths);
+    }
+
+    async rm(repository: Repository, uri: string | string[]): Promise<void> {
```
I am pretty sure we could change the API in the future if we have to or just want to. I do not see this as a blocking issue now, do you? Maybe I am just missing something.

hexa00 replied:
Well givin that it's just the name of a function I don't see the harm in changing that now?

kittaakos replied:
I have changed it for now. We need to find the proper method name in the future, which is an eventual Git command.

hexa00 replied:
OK, I don't think it has to be a git command otherwise it is git reset...

kittaakos replied:
>  I don't think it has to be a git command

You might have accidentally overlooked one of my comments; too long PR. We would like to align the method names (with the functionality) with the Git commands. I hope this clarifies.

kittaakos commented on packages/git/src/common/model.ts:
```diff
@@ -0,0 +1,111 @@
+    'New',
+    'Modified',
+    'Deleted',
+    'Renamed',
```
Same as [here](https://github.com/theia-ide/theia/pull/392#discussion_r142915482). If the result can be `Renamed` according to Git, then we have to support `Renamed`. We did not invent these on our own.

hexa00 replied:
OK thanks that's what I wanted know.

hexa00 commented:
All that is left in my view now is to test on windows, I should be able to do that today.
Please be patient it's 8hAM here...

hexa00 commented:
@kittaakos I just tested on windows:

 - There's an issue with the diff editor, it shows the left side as empty and the right side with the whole file instead of the chagnes only. (this worked on linux)

Otherwise it seems to work.

kittaakos commented:
> There's an issue with the diff editor, it shows the left side as empty and the right side with the whole file instead of the changes only. (this worked on linux)

That is unfortunate. I will try to write some core Git tests to check whether it is due to Git or Monaco.

svenefftinge commented:
Sounds to me like our 'git'-uri based `Resource` is not working on windows.

kittaakos commented:
I will give it another try on Windows, but I was unable to run the application with my configuration last week. I will check.

kittaakos commented:
> Sounds to me like our 'git'-uri based Resource is not working on windows.

Indeed the [git-resource seems to be broken](https://github.com/theia-ide/theia/blob/GH-165/packages/git/src/browser/git-resource.ts#L50). You cannot just compare a URI as a string with a file path. The URI can be encoded; file path will not be.

I will take care of this.

kittaakos commented:
I have found and fix the resource bug, but I think there is another in `git show`. I need to look into that. Plus the Git API had another issue. `show` returns with a `Promise<String>` instead of a `Promise<string>`. I try my best to commit all the changes together.

kittaakos commented:
There is another problem with the repositories lookup, paths are sometimes normalized, sometimes not:
```bash
1 [ 'file:///c%3A/Users/KITTAA~1/AppData/Local/Temp/discovery-test-211796-8240-vm7sah.wow0b/BASE',
  'file:///c%3A/Users/KITTAA~1/AppData/Local/Temp/discovery-test-211796-8240-vm7sah.wow0b/BASE/A',
  'file:///c%3A/Users/KITTAA~1/AppData/Local/Temp/discovery-test-211796-8240-vm7sah.wow0b/BASE/B',
  'file:///c%3A/Users/KITTAA~1/AppData/Local/Temp/discovery-test-211796-8240-vm7sah.wow0b/BASE/C' ]
2 file:///c%3A/Users/kittaakos/AppData/Local/Temp/discovery-test-211796-8240-vm7sah.wow0b/BASE

```

kittaakos commented:
I have fixed the diff editor issue. @hexa00, please review if you have time. Thank you!

cc: @svenefftinge
